File: core/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.py
from typing import List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def levelOrder(root: TreeNode) -> List[List[int]]:
    if not root:
        return []
    queue = deque([root])
    res = []
    while queue:
        length = len(queue)
        level = []
        for _ in range(length):
            node = queue.popleft()
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
            level.append(node.val)
        res.append(level)

    return res

File: core/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.java
package com.dbc;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeLevelOrderTraversal {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>() {{
            add(root);
        }};

        List<List<Integer>> res = new ArrayList();
        while (!queue.isEmpty()) {
            int length = queue.size();
            List<Integer> temp = new ArrayList();
            for (int i = 0; i < length; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
                temp.add(node.val);
            }
            res.add(temp);
        }
        return res;
    }
}
